File: exercises/arrays/selectAboveMean/selectAboveMean.js
/**
 * Given an array of numbers, returns a new array consisting of every
 * number in the input array strictly greater than the mean.
 *
 * Return an empty array if no such numbers exist.
 *
 * @example
 * selectAboveMean([1, 1, 1]); // => []
 * selectAboveMean([-10, -10, 0, 10]); // => [0, 10]
 * selectAboveMean([1, 2, 3, 4, 5]); // => [4, 5]
 * selectAboveMean([100, 0, 100, 200, 100, 0]); // => [100, 100, 200, 100]
 *
 * @param {number[]} array - An array of numbers
 * @returns {number[]} A new array consisting of all numbers in the input array
 *   greater than the mean.
 */
function selectAboveMean(array) {
  function getMean(array) {
    let sum = 0;
    for (let num of array){
      sum += num;
    }
    return sum/array.length;
  }
  const mean = getMean(array);

  const newArray = [];
  for (let element of array){
    if (element > mean){
      newArray.push(element);
    }
  }
  return newArray;
}

function outputEqualsExpected(output,expected){
  if (output.length !== expected.length){
    return false;
  }
  
  for (let i = 0; i < output.length; i++){
    if (output[i] !== expected[i]){
      return false;
    }
  }

  return true;
}

if (require.main === module) {
  console.log('Running sanity checks for selectAboveMean:');

  // Add your own sanity checks here.
  // How else will you be sure your code does what you think it does?
  console.log(outputEqualsExpected(selectAboveMean([1,2,3,4]), [3,4]));
  console.log(outputEqualsExpected(selectAboveMean([1,2,3,100]), [100]));
  

}

module.exports = selectAboveMean;

File: exercises/arrays/selectAboveMean/selectAboveMean.py
# Import reduce function
from functools import reduce



def selectAboveMean(numberList):
  # Get mean of list
  sum = reduce((lambda x,y: x + y), numberList)
  mean = sum/len(numberList)

  result = list(filter((lambda x: x > mean), numberList))
  return result

if __name__ == "__main__":
  print('Running sanity checks for selectAboveMean')

  tests = [
    {
      'list':[1,2,3,4],
      'expected': [3,4]
    },
    {
      'list': [1,2,3,100],
      'expected': [100]
    }
  ]
  for test in tests:
    print(selectAboveMean(test['list']) == test['expected'])
